### DISCUSSION TITLE
Choirs using Jamulus

### DISCUSSION THREAD

**Discussion by hoffie** [jamulussoftware/jamulus#1456] [General]
It seems that more and more choirs started using Jamulus.

Without any specific goal, I'm asking: Who are you? What are your experiences? :)
(If you don't feel comfortable posting names or exact locations, just mentioning numbers, obstacles or other relevant details would be interesting :))

I regularly sing with both mixed and mens choirs with 10-25 people from Bamberg, Bavaria, Germany. We currently don't practice for anything specific, we just enjoy singing together, both previously known repertoire and new stuff. Initial setup (mostly with internal microphones and ASIO4ALL) was a huge hassle, but once it had been set up, it works fine.

(It think there may have been similar groups or threads on Facebook)

Cc'ing some people who I think may have choir-ties: @DominikSchaller @WolfganP @ann0see @cwerling @DetlefHennings @kezzy1966 @menzels @dingodoppelt

**Comment by ann0see** [comment]
Interesting post ;-). 

The choir I sing in is a German boys choir. We're singing with Behringer UPhoria UMC202HD Interfaces since I really wanted to skip ASIO4ALL support hell. I still think this was one of the best choices we made. Of course interfaces cost money, but a decent interface isn't too expensive. 

We've already streamed a concert to our church and Zoom. 

Unfortunately we're not that many singers yet (maybe 20) but we're slowly onboardig more and more people. Many told me that they were really fed up with video conferencing singing and are really enjoy singing with Jamulus.

**Reply by ghost** [reply]
Hi, My friend bought a UMC202HD sound Card, and use it with a MAC. He can hear everything, but his sound, microfon and the Bass is not coming through. Any suggestion or does anyone had a similar issue. Br Leif

**Reply by ann0see** [reply]
Not sure about any solution I'm afraid. You might ask Behringer?

Can you open another discussion on that please?

**Reply by wolli** [reply]
Hier ist noch ein Alter langjähriger Sänger,
der sich gerade per Apple versucht in diese tolle Software ein zu arbeiten.

übrigens, steht die Gründung eines Shanty Chores kurz bevor sobald es die total verkorkste Politik zulässt.

LG in die Runde aus Limburg/Lahn
Wolle

**Reply by wolli** [reply]
Hm zusammen,
scheint keine einfache einheitliche und kostengünstige Lösung zu geben.

Schade und LG
Wolle

**Comment by dkmj** [comment]
I'm in a Swedish chamber choir, based north of Stockholm. We are just about to start, after a week or three of getting members to understand tech requirements and investing in microphones and head phones. first actual rehersal a few days from now. we'll see how it goes. 

im running the server at home. have yet to understand performance constraints on the server side. Using a MacMini 2021 cor i7 32Gb Ram.

**Reply by dkmj** [reply]
Also installed a server on a Raspberry Pi 4 4gb, running this headless in parallell with my mac server. different ports. I am trying to keep track of jamulus server performance. it seems there is plenty CPU and memory headroom on both machines. how would a performance bottleneck manifest? worse audio quality? pops, cracks? thown users? we are in the early stages and surely most problems are caused by non techie users. but. if when we get to server bottlenecks, how would I be able to tell? 

is there any way of loading / testing the server to capacity? 20 user max on the Mac server, 35 in the Raspberry Pi 4, for now. 

thanks, this is fun!

**Reply by dcorson-ticino-com** [reply]
My experience has been that at some point it is no longer possible to connect to the server or that the connection takes a very long time.
I do not remember if there are tell-tale audio artifacts pointing to server overload.
You can use top or something similar to see the processor loading.
In any case, nearly all sound quality problems are caused by clients and their audio hardware and have nothing to do with the server.

**Reply by menzels** [reply]
I experienced high latency spikes and dropouts when the server was running on near full load. 
Most obvious sign was the ping going to infinity and also what @dcorson-ticino-com described.

**Reply by WolfganP** [reply]
I think that Jamulus' usage on large groups pushes it to (some) limit that we don't have many tools to troubleshoot properly. I opened an instrumentation focused discussion at https://github.com/jamulussoftware/jamulus/discussions/1547 if someone is interested to participate at.

**Reply by gene96817** [reply]
On my Raspberry Pi 4, performance is good with 15 singers (overall latency ~40-50ms). Going to 16 - 17 singers, the overall latency jumps to >500ms. The client UI stops working, additional singers cannot connect, the sound becomes terrible. The CPU loading is 35%-40% on each of the RPi4 four cores.

**Reply by gene96817** [reply]
> I experienced high latency spikes and dropouts when the server was running on near full load.

What CPU loading is considered full load? My RPi4 behaves as if it is fully loaded when the CPU load approaches 40%. That seems to be a very low loading level.

**Reply by dcorson-ticino-com** [reply]
>My RPi4 behaves as if it is fully loaded when the CPU load approaches 40%

Is the cpu clock management working?  On my Odroid-N2 (also an ARM processor) at about 40% it does a frequency step.  Even at 25 clients it is not yet running at full frequency and the capability of the processor doesn't test at that much more than a Pi4.
My calculations say at 10 users the Odroid is running at about 22% load.  I think the Pi4 should be able to do much better than 15 users.

**Reply by menzels** [reply]
on my cloud instance these effects started at about 90% per core cpu load.
i noticed though that top or htop sometimes displays misleading values, it sometimes displays nearly no cpu load overall. but if i look at the individual processes, i see that there is significant load.

**Reply by gene96817** [reply]
> Is the cpu clock management working? 

What are you suggesting? What feature of Ubuntu manages the "CPU clock management"? Are you suggesting that the CPU needs to be overclocked? If so, at what speed? I assume for maximum reliability, I would run the CPU at the max rated clock speed.

**Reply by dcorson-ticino-com** [reply]
Hi Gene,
I am not an expert for Ubuntu or Linux, but I have for the Odroid the possibility to manage the processor speed to save power and lower the heat production.  At idle the processor is running at 1Ghz, when the load goes up to about 40% it notches up the processor speed in steps 1.2G, 1.398G, 1.608G etc. up to the rated speed of 1.8Ghz.  I am not overclocking at all and this appears to me to be a standard part of Ubuntu in the lib cpufrequtils.

**Comment by chrisrimple** [comment]
You may also want to ask your question in the Facebook groups [Jamulus Choral Community](https://www.facebook.com/groups/jamuluschoralcommunity) and/or [Jamulus (official group)](https://www.facebook.com/search/top?q=jamulus%20(official%20group)).

**Comment by mhilpolt** [comment]
I'm using Jamulus singlemixserver since beginning of February with our chorus in Regensburg/Germany with regularly around 35 participants. We are using the JackTrip client image, which is quite easy to use for all singers, the master mix is running on a Mac, the director uses a Jambox image. So we solely use Raspbery Pi with a simple USB-Audiointerface and electret neckband microphones. With that common approach, we get a well-balanced sound with an overall delay of around 40-60 ms. It works great!!!
The only feature I hardly missed is a real "Solo" feature for the master mix, but @DominikSchaller told me that @cwerling already implemented that feature: EXCELLENT! Will now test it ...
Maybe some experience regarding my server: I've tested Google Cloud and 1Fire VM's but got some loss in audio quality when connecting more than 8 people. With IP-Projects dedicated server (lowest profile, 25,- Euro/Month), it works fine - lets say up until 50 connections (singlemix, buffer 128, high quality, mono-in/stereo-out)

**Comment by cwerling** [comment]
Heyho choir colleagues! I am with the mixed choir [Charite Chor Berlin](https://www.charitechorberlin.de/) and we currently rehearse the [Requiem by Fauré](https://en.wikipedia.org/wiki/Requiem_(Faur%C3%A9)) over Jamulus. Happy to share our current modus operandi:

We rehearse with Jamulus since the beginning of this year, using a variety of setups. People who didn't have a mic bought a t-bone USB microphone as well as Ethernet cables (and adapters where needed).

Our most valuable addition to Jamulus are the free [Jitsi](https://meet.jit.si/) calls that we use to stream our [conductor](https://www.adrianemans.de/)'s video feed to us. Before having this we practiced using an acoustic click, but the conducting provides just so much more musical information and spares us the annoying click sound. In our experience, Jitsi has a far lower latency than e.g. Zoom. Some of us use a separate device for the video feed, since Jitsi in the browser can be quite resource hungry. With low-bandwidth DSL connections it can be worth it to even use mobile data for the best quality video. With setting the video quality to low in Jitsi (which is still pretty good), I used around 200 MB of mobile data for a 2-hour rehearsal.

We still test my [`--mastermix` PR](https://github.com/jamulussoftware/jamulus/pull/1381), but it relieves most of us from interacting with Jamulus at all during the rehearsal. We do a soundcheck at the beginning of the rehearsal where our conductor sets everybody's gain one voice group after the other (sopranos, altos, tenors, bass). Everybody is still able to adjust our conductor's gain as well as their own to their own liking, although we recommend everybody to mute themselves from their mix. (This seems to be an unpopular choice, but having overall latencies of around 60-80ms it works very well for us and we forget about delays pretty much).

We run the Jamulus Server on an Amazon EC2 t3.xlarge using Docker. I started using an AWS Lambda function along AWS CloudWatch to automatically start and stop the EC2 instance for our rehearsals on a weekly basis. Costs are about 5 USD per rehearsal.

**Reply by WolfganP** [reply]
> Our most valuable addition to Jamulus are the free [Jitsi](https://meet.jit.si/) calls that we use to stream our [conductor](https://www.adrianemans.de/)'s video feed to us. Before having this we practiced using an acoustic click, but the conducting provides just so much more musical information and spares us the annoying click sound. In our experience, Jitsi has a far lower latency than e.g. Zoom. Some of us use a separate device for the video feed, since Jitsi in the browser can be quite resource hungry. With low-bandwidth DSL connections it can be worth it to even use mobile data for the best quality video. With setting the video quality to low in Jitsi (which is still pretty good), I used around 200 MB of mobile data for a 2-hour rehearsal.

If you only need a one way/broadcast video stream for the conductor and all audio managed by **jamulus**, maybe switching to youtube or twitch private streams will work better for your group as you'll have access to mature multiplatform clients (or even their browser optimized interfaces) and very low traffic needed in low quality settings.

**Reply by DominikSchaller** [reply]
I don't think Youtube comes close to the required latency.

**Reply by WolfganP** [reply]
> I don't think Youtube comes close to the required latency.

You're maybe right, the stream base delay by going thru a 3rd party service/infrastructure like YT or Twitch may be too much for online jamming reaction to the conductor.

It's not that I'm again Jitsi (I love it and in fact I hosted a private instance for my friends & family online events during quarantine), but I was wondering if your use case can be optimized going from a conference (two way) solution to a broadcast setup (one way director feed, saves resources on server and all clients attached as no feedback stream is needed).

As Jitsi uses WebRTC, it may be interesting to try a broadcasting solution based on the same tech and see if the overall experience improves. https://www.webrtc-experiment.com/webrtc-broadcasting/ or https://www.webrtc-experiment.com/broadcast/ (https://github.com/muaz-khan/WebRTC-Experiment/tree/master/broadcast) seems to be cool experiments to try. I'll see if I can make some time and test them...

**Reply by DominikSchaller** [reply]
You're absolutely right. Youtube's ultra low latency mode is at 3 seconds, so not usable for us. But a custom broadcasting solution would be ideal.

**Reply by WolfganP** [reply]
This seems promising as well https://github.com/GRVYDEV/Project-Lightspeed (a self contained OBS -> FTL -> WebRTC live streaming server)

**Reply by Hk1020** [reply]
Nextcloud Talk? Just an idea, could be rubbish.

**Reply by j-santander** [reply]
Could you provide further details on your jitsi setup? Are you using public jitsi servers or are you connecting to your own private instance?

**Reply by cwerling** [reply]
> Could you provide further details on your jitsi setup? Are you using public jitsi servers or are you connecting to your own private instance?

Surprisingly, the meet.jit.si servers worked just as well as a custom Jitsi service running in the same server we run Jamulus on. Since meet.jit.si’s free tier had everything we need (and better server maintenance etc.), we use that now. 

Important Jitsi settings:
- The moderator should set everybody joins muted in the settings
- People streaming should lower their quality settings to low to avoid stealing Jamulus the local DSL’s bandwidth
- People should ideally join completely without sound (in and out) using the little arrow on the joining screen [0]
- In case someone still accidentally unmutes and disturbs the Jamulus Jam, the moderator can mute everybody at once in the settings

[0] I wonder if it’s possible to set this (and maybe even the quality settings) globally for everyone joining a specific Jitsi meeting. I saw there are some hidden settings that can be set through GET parameters, but haven’t looked into it.

**Reply by ann0see** [reply]
I found these options (not yet tested): `https://meet.jit.si/JamulusgChoirRehearsal5289#config.prejoinPageEnabled=false&disableInviteFunctions=true&config.startSilent=true&config.startWithVideoMuted=true&config.disableAudioLevels=true&config.resolution=180&config.constraints.video.aspectRatio=16/9&config.constraints.video.height.ideal=200&config.constraints.video.height.max=200&config.constraints.video.height.min=200&config.p2p.enabled=false&config.disableAGC=true&config.disableHPF=true&config.disableNS=true&config.enableLipSync=false&config.enableLayerSuspension=true&config.disableE2EE=true&config.stereo=true&config.enableNoisyMicDetection=false&config.disableAP=true&config.requireDisplayName=true `

**Reply by hoffie** [reply]
Thanks for these Jitsi tips!

> I wonder if it’s possible to set this (and maybe even the quality settings) globally for everyone joining a specific Jitsi meeting. I saw there are some hidden settings that can be set through GET parameters, but haven’t looked into it.

I have searched for options to disable microphone permission requests and only found hints that this isn't possible (yet). I have come up with a quick hack to set useful low-latency options (for browsers only) by placing them in `web/custom-config.js` of my self-hosted Jitsi instance (docker-compose-based).

We have tried using Jamulus + Jitsi at the same time once and Jamulus experience quickly degraded. I'm not sure if it was due to video encoding/decoding performance on clients, due to bandwidth (unlikely with the limits I had configured) or due to the fact that Jitsi was running next to the Jamulus server.

```javascript
if (window.location.hash.indexOf('jamulus') >= 0) {
        console.log("disabling audio via getUserMedia monkey-patching");
        navigator.mediaDevices.getUserMedia_orig = navigator.mediaDevices.getUserMedia;
        navigator.mediaDevices.getUserMedia = async function(constraints) {
                if ('video' in constraints && constraints['video']) {
                        constraints['audio'] = false;
                } else {
                        return false;
                }
                console.log("overriden constraints: ", constraints);
                return await navigator.mediaDevices.getUserMedia_orig(constraints);
        };

        config.resolution = 320;
        config.constraints.video.height = { ideal: 320, max: 720, min: 120 };
        config.constraints.video.frameRate = { ideal: 10, max: 30, min: 1 };
        config.constraints.video.width = undefined;
        config.opusMaxAverageBitrate = 0;
        config.enableNoAudioDetection = false;
        config.enableNoisyMicDetection = false;
        config.startAudioMuted = 1;
        config.startWithAudioMuted = false;
        config.videoQuality.maxBitratesVideo = {
                low: 10000,
                standard: 10000,
                high: 10000
        };
        config.maxFullResolutionParticipants = 0;
        config.disableTileView = false;
}
```
Then, use https://meet.example.org/RegularMeetingName#jamulus as the link for participants. I'm mainly posting this here for the microphone permission hack. As posted above I'm not satisfied with the overall settings. Requesting really low frame rates hasn't worked for me either (Linux/Firefox).

**Comment by mhilpolt** [comment]
So, today our rehearsal took place with mastermix-branch. I've found the code passage to remove the individual channel settings so that it was working in my setup. In general, everything worked fine, no crash, CPU-Load around 65%, Solo feature worked fine for me but:
- there were more noise and cracklings in the sound
- the sound seemed to be not that direct and somehow delayed as I was used to with singlemix
- I tried to find a connection with a higher noise rate by using the Solo feature, but not any single channels had notable interferences, but the mixed sound.

Not really objective. After some time, we switched to the well-tested singlemix-branch. Even if some interferences were still preset, it seemed to be better and more direct.
My comments:
1. multithreading might be a reason for interferences
2. CPU-load might also be a reason for interferences
3. With my simulation with the same amount of connection, I did not observe that cracklings -> it seemed to be based on the connections from different computers (my test setup creates many clients from one machine) even if I got the same CPU-Load.

Question: Can anybody think of a source of disturbances during the mix-process? As I stated above, using the Solo-feature, I was not able to find single channels which produces much interferences, but in the mix they were present.
How can I measure the quality of a connection/mix? If I change some parameters, it would be very helpful to get a measurement-value (like a harmonic factor) to rate the changes... Has someone experience here?

**Comment by brb214** [comment]
I am in a choir that currently only uses Zoom. I downloaded Jamulus and installed the ASIO driver. When I connect to a server, I don't hear anything. Is this because I am the only person on the server? I can see the green light move in the left and right input bars. I read the troubleshooting guide, and logged into a server with a larger delay and could not hear any delayed sounds. I am using a standard jack headphone from Jamba. I see the ASIO software icon at the bottom of my window, and am closing Skype everytime I try to log onto a server. Can anyone tell me what I could be doing wrong? I am running Windows 10.

**Reply by ann0see** [reply]
Hi! This seems to sound like a wrong configuration of the output device in ASIO4ALL (the ASIO driver you’re probably using). I assume the automatic selection of the output is wrong. See https://jamulus.io/wiki/Installation-for-Windows#asio
on how to configure it and https://jamulus.io/kb/2021/03/20/ASIO4ALL-Examples.html for working example configurations. If this doesn’t work, please open another post to raise visibility to your question.

I must warn you that ASIO4ALL and it’s configuration is probably the hardest thing to do for a choir (it often doesn’t work out of the box correctly and needs to be properly configured). Also internal sound devices are noisy and far from perfect. That’s why we’re using Interfaces.

**Reply by brb214** [reply]
Good morning! Thank you for the tip. I looked into the ASIO4ALL settings this morning, and now see my microphone and headset as listed devices. I don't know what I did differently, but based on your tip made sure to unselect the other devices in the list and could log in to a server and hear myself. 

I appreciate the warning and will pass it on.

**Reply by ann0see** [reply]
Probably you plugged it in after you opened jamulus. Therefore you would have needed to either reset ASIO4ALL, restart Jamulus or reload the sound system (click on the ASIO4ALL driver in the driver selection menu even if it was already selected).

**Comment by j-santander** [comment]
We're a SATB Choir ([Polytechnic University of Madrid Choir](https://www.coro-upm.es)) and we've been doing jamulus rehearsals once a week since October 2020, supplementing the one-way Zoom meetings that we've been using for explanation, instructions and analysis of the pieces we've prepared. 

The biggest hurdles have been technical:
- People need to have a PC (the times of the PC as a house appliance are over), moderately up-to-date and where they can install software (corporate laptops might not allow to install unapproved software).
- People need to be slightly tech-savy, at least to know how to install software.
- ... and then starts the problems with the audio devices (and ASIO4ALL in Windows), learning that there might be many of them, that USB headphones create their own device.... that there're input gains to consider (a setting hidden under many layers of Windows configuration).
- People must use headphones and microphone. Specially at the beginning it was difficult to convince everybody.... 
  - and still, they should be decent enough (sometimes cheap mobile earphones work, sometimes they have a dynamic range tailored for conversation making it bad for singing). For some people, not sure why (too loud headphone too close to the microphone?), the output still feeds into the input
  - those that already had or invested in a youtuber/professional-microphone, those seem to be very sensitive to the distance from the mouth, something easy to forget while singing.
- WiFi is usually not good enough (but I suspect people still connect through it... as they don't have other options).

We've been singing using midi bases and metronome. Recently we've added a simultaneous (silent) Zoom meeting because it was nice to see other people faces, but the delay was not good enough to follow a conductor.

Before Christmas we tried to rehearse Mozart's Coronation Mass. It didn't go very well:
- Probably we were still learning and there were many people with bad connections/headphones.
- Coronation Mass is not that difficult, but required some pre-rehearsal work that many people was not doing.

After Christmas we changed things:
- We move to *a cappella* works, more vertical.
- We divided the choir in rehearsals groups of aprox 16 persons.
- We settled on server running on a VPS with 2 CPUs

With that setup things have been working much better. Our choirmaster sometimes can remove the base and metronome completely and it really sounds like singing.

My personal impression is that overall results is highly critical with someone having a bad set up (or bad knowledge). The risk of having a *misbehaved* user increases with the number of people, that's why we end up with small groups. You could mute them, but then you need to be continously fiddling with the controls instead of singing.

We do usually a roll call to adjust the levels... but:
- People talking is not the same as people singing... so ask them to sing, not just "one, two, three, testing...."
- Many people don't know how to control the mic input gain (usually the will confuse the input gain with the output volume and then they will complain that they don't hear anybody). They usually end up moving the microphone closer (and forget about it later and move away)... or everyone have to use the fader on them.

Many times there's this *watery* sound that I guess comes from network underruns (bad connections? wifi connections?) 

We recently trialed Sonobus: the consensus was that it sounded much better, but for me it was technical nightmare as Sonobus is p2p and we have to resolve the N*(N-1) connectivity problems (firewalls, port-forwarding, router config...)

**Comment by cwerling** [comment]
Yesterday we learned that running the Jamulus server on my rather slow MacBook through my home DSL in Berlin (50,000 Mbps) produces better sound/performance for our Berlin-based, 20-people choir than using an Amazon EC2 t3.xlarge hosted in Frankfurt. I find that very counterintuitive given both the CPU and network performance the EC2 machine should have.

It can't be just the slightly lower ping to the shorter distances covered (pings were fine in both setups). Is there a chance this has to do with the way the virtualized EC2 instances schedules tasks?

<img width="1153" alt="Screenshot 2021-04-22 at 11 14 05" src="https://user-images.githubusercontent.com/8776698/115688904-dcff4380-a35b-11eb-9171-ff9c6fa04d14.png">

**Reply by menzels** [reply]
I use Hetzner for cloud hosting and they have "dedicated" cloud instances that i found work with better consistency.
This looks like a scheduling issue between instances, but i did not investigate further.

**Reply by gene96817** [reply]
Scheduling is a good issue to investigate. The fundamentals of a cloud service is to maximize use of all the resources through sharing among multiple customers. Different services have slightly different sharing approaches. Sharing means there are differences between your steady state performance and the burst performance. Alternatively, think of this as the mix between burst times and fallow (wait) times. 

My general inclination is that steady state performance will yield better results than bursty processing.

**Comment by Andrew999999** [comment]
Has anyone tried using the **android** jamulus client with choir members?  It requires a wired earpiece (not bluetooth) - and ideally a wired ethernet connection (e.g. via USB-on-the-go connection - not all phones will support this but many do).  Smartphones mics are tuned for voice - so my tests with a musical instruments were not good - but I'm wondering if this would work ok for singing?
